pkg/dynconf/dyn_settings.go

```diff
@@ -50,46 +52,63 @@ func readSettings(commandFlagSet *pflag.FlagSet, withContext bool) *DynSettings
 		if err != nil {
 			log.Fatalf("could not find config file '%s'\n", configFileName)
 		}
-		err = yaml.Unmarshal(source, allParams)
+		tempConfig := &map[interface{}]interface{}{}
+		err = yaml.Unmarshal(source, tempConfig)
 		if err != nil {
 			log.Fatalf("invalid syntax in config file '%s': %s\n", configFileName, err)
 		}
+		validateProps(*tempConfig, configFileName)
+		if err = subEnvVars(tempConfig); err != nil {
+			log.Fatalf("bad substitution in '%s': %s\n", configFileName, err)
+		}
+		if configParams, err = convertMap(*tempConfig); err != nil {
+			log.Fatalf("invalid syntax in config file '%s': %s\n", configFileName, err)
+		}
 	}
 
-	// Merge before validation and env substitution since it might not be needed due to context.
+	// Read context, if it should be included.
 	if withContext {
-		mergeContext(allParams, contextName)
+		contextHandler := NewContextHandler()
+		if contextName == "" {
+			contextName = contextHandler.Current
+		}
+		context := contextHandler.Get(contextName)
+		contextParams = map[string]interface{}(context)
 	}
 
-	validateProps(*allParams, configFileName)
-	err = subEnvVars(allParams)
-	if err != nil {
-		log.Fatalf("bad substitution in '%s': %s\n", configFileName, err)
+	// Check for overlap in config and context.
+	for k := range contextParams {
+		// Headers from context will be merged with that of config, so no conflict.
+		if k == "headers" {
+			continue
+		}
+		if _, ok := configParams[k]; ok {
+			log.Warnf("Property '%s' exists in both current context and config, using property from config\n", k)
+		}
 	}
 
+	// Read command-line parameters
 	commandFlagSet.Visit(func(flag *pflag.Flag) {
 		value := getFlagValue(commandFlagSet, flag)
-		(*allParams)[flag.Name] = value
-		(*commandLineParams)[flag.Name] = value
+		commandLineParams[flag.Name] = value
 	})
 
+	// Read default values
 	commandFlagSet.VisitAll(func(flag *pflag.Flag) {
-		if (*allParams)[flag.Name] == nil {
+		if _, ok := commandLineParams[flag.Name]; !ok {
```

Comment:
No real point in doing this check if we modify the getter down below (see comment there). Otherwise we need to check if the flag is specified in commandLineParams OR configParams or contextParams here

Reply:
Fixed now!